### PR TITLE
Adding hint text for a query SQL

### PR DESCRIPTION
**Pull request by xJoeWoo**
Adds the ability to mutate the query SQL with hint text. Besides, adjusted `AbstractQuery.copyTo` function to public visibility and open to make extending the `Query` class easier.

### CODE REVIEW

**Comment by xJoeWoo** [comment]
This solution has its limitation: only supports `Query` statement. If the hint needs to be added to all statements, for example, used for tracking, another mechanism should be added to `Statement` class, or the `StatementInterceptor` class.

I recommend making `StatementInterceptor` more powerful, like intercept then editing the statement, or even the final SQL string. This makes updating the `updateTime` column(which should be updated every time the row changes) much much simpler, no need to assign `updateTime` column everywhere, and no worry about forgeting to update this column. 

```
interface StatementInterceptor {

    // add this function, call in the first line of org.jetbrains.exposed.sql.statements.Statement#executeIn
    fun beforeArgument(transaction: Transaction, statement: Statement<*>) {}

    // even mutating the SQL, useful for adding tracking hints
    fun beforeStatementPrepared(transaction: Transaction, sql: String): String = sql

}
```
``` 
class UpdateTimeColumnType : ColumnType<LocalDateTime>()

object UpdateTimeGlobalStatementInterceptor : GlobalStatementInterceptor {
    override fun beforeArgument(transaction: Transaction, statement: Statement<*>) {
        val updateStatement = statement as? UpdateStatement ?: return
        val table = updateStatement.targetsSet as? Table ?: return
        val column = table.columns.firstOrNull { it.columnType is UpdateTimeColumnType } as? Column<LocalDateTime> ?: return

        updateStatement[column] = LocalDateTime.now()
    }
}
```


```

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/AbstractQuery.kt [hunk 1]
```diff
@@ -31,7 +31,12 @@ abstract class AbstractQuery<T : AbstractQuery<T>>(
     /** The set of columns on which a query should be executed, contained by a [ColumnSet]. */
     abstract val set: FieldSet
 
-    protected fun copyTo(other: AbstractQuery<T>) {
+    protected abstract fun copyInstance(): T
+
+    /** Creates a new query instance using all stored properties of this query. */
+    override fun copy(): T = copyInstance().apply { copyTo(this) }
+
+    open fun copyTo(other: T) {
```
Comment (bog-walk) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/AbstractQuery.kt:39]: Could you please explain the reason for introducing `copyInstance()` and overriding `SizedIterable.copy()`?
Why not just make `copyTo()` open and leave everything else the same?

File: exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/dml/SelectTests.kt [hunk 2]
```diff
@@ -564,4 +564,26 @@ class SelectTests : DatabaseTestsBase() {
             assertEquals(0, stringTable.selectAll().where { stringTable.name eq veryLongString }.count())
         }
     }
+
+    @Test
+    fun `test hint select query`() {
```
Comment (bog-walk) [inline exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/dml/SelectTests.kt:569]: Please use camelCase test naming and add the new test as a `@sample()` to the new `fun hint()`.

Please see [contribution style guide](https://jetbrains.github.io/Exposed/contributing.html#style-guides) for more details.

**Review by bog-walk — CHANGES_REQUESTED** [review]
Thanks for putting together this feature @xJoeWoo . Please see comments in the files.

Also, please take a look at the [contribution guidelines](https://jetbrains.github.io/Exposed/contributing.html#pull-requests), particularly steps 9 and 10, to pass the `detekt` and `apiCheck` tasks.

**Comment by bog-walk** [comment]
Hi @xJoeWoo It would be great to include this feature in the upcoming release. So I will rebase and address the requested changes above on this branch next week if you are not able to. And the additional adjustment to the visibility modifier to make the `Query` class more extensible will be done in a separate dedicated PR (so I can add relevant documentation and tests).

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Query.kt [hunk 3]
```diff
@@ -373,6 +420,15 @@ open class Query(override var set: FieldSet, where: Op<Boolean>?) : AbstractQuer
...
     }
+
+    /** Represents the position at which an SQL comment will be added in a `SELECT` query. */
+    enum class CommentPosition {
+        /** The start of the query, before the keyword `SELECT`. */
+        FRONT,
+
+        /** The end of the query, after all clauses. */
+        BACK
+    }
```
Comment (bog-walk) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Query.kt:431]: I considered changing this to `START` and `END` but couldn't find many similar examples and figured it was still clear.

**Comment by bog-walk** [comment]
Changes made:

- Rebase and add `apiDump`
- Remove changes involving extension of custom query classes (now in PR #2173 )
- Switch naming to `comment()` to align with the standard sql
- Add `adjustComments()` that allows existing query comments to be removed/updated
- Adjust test accordingly

**Review by bog-walk — APPROVED** [review]

**Review by e5l — APPROVED** [review]

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Query.kt (hunk 3 above)
Comment (fokkoru) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Query.kt:431]: @e5l  @bog-walk Could we add an option to insert a comment after the SELECT keyword to support [Optimizer Hints](https://dev.mysql.com/doc/refman/8.4/en/optimizer-hints.html)?

Comment (bog-walk) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Query.kt:431]: Hi @fokkoru Yup, that's a popular and useful case, so I could add something like `CommentPosition.AFTER_SELECT`.
In the meantime though, if you need it, here's a custom query for inspiration:
```kt
fun Query.optimizerHint(hint: String) = OptimizerHintQuery(this, hint)

class OptimizerHintQuery(val source: Query, val optimizerHint: String) : Query(source.set, source.where) {
    init {
        source.copyTo(this)
    }

    override fun prepareSQL(builder: QueryBuilder): String {
        val originalSql = super.prepareSQL(builder)
        return originalSql.replace("SELECT ", "SELECT /*$optimizerHint*/ ")
    }

    override fun copy(): OptimizerHintQuery = OptimizerHintQuery(source.copy(), optimizerHint).also { copy ->
        copyTo(copy)
    }
}
```

**Pull request merged**